### PR TITLE
WIP: support worker and metricsCollector template spec

### PR DESCRIPTION
**Pull request by hougangliu**
Fixes:  #349

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/katib/375)

### CODE REVIEW

File: pkg/controller/studyjob/manifest_parser.go [hunk 1]
```diff
@@ -28,30 +27,50 @@ import (
...
-		if err != nil {
-			return "", nil, err
-		}
-		if wt, ok := wtl[wPath]; !ok {
-			return "", nil, fmt.Errorf("No worker template name %s", wPath)
-		} else {
-			wtp, err = template.New("Worker").Parse(wt)
+		tName, tNamespace, tPath := getDefaultTemplateSpec()
+		if goTemplate.TemplateSpec != nil {
+			tName = goTemplate.TemplateSpec.ConfigMapName
```
Comment (YujiOshima) [inline pkg/controller/studyjob/manifest_parser.go]: When the field of TemplateSpec is emply, we should use the default value.
For example, we can omit the namespace of ConfigMap when we want to use the same namespace as the studyjobcontroller.

File: pkg/api/operators/apis/studyjob/v1alpha1/studyjob_types.go [hunk 2]
```diff
@@ -114,19 +114,25 @@ const (
 	OptimizationTypeMaximize OptimizationType = "maximize"
 )
 
-type GoTemplate struct {
+type TemplateSpec struct {
```
Comment (YujiOshima) [inline pkg/api/operators/apis/studyjob/v1alpha1/studyjob_types.go]: This extension of studyjob API would be for v1alpha2? https://github.com/kubeflow/katib/issues/370

**Comment by YujiOshima** [comment]
This issue is listed v1alpha2 API https://github.com/kubeflow/katib/issues/370 .
@johnugeorge Should we make a branch for v1alpha2 and this will be merged to the branch?

**Comment by johnugeorge** [comment]
Yes. I would suggest that only common changes(that won't break the current config) go into v1alpha1 at this point. @richardsliu is working to get v1alpha2 structure. Once it is merged, we can get it in v1alpha2

In general,  I feel that we are moving away from the data scientist point of view, instead thinking from a kubernetes developer perspective.   Currently, a data scientist will find it extremely difficult to write a katib study job config correctly without good kubernetes experience. For eg: Currently in the studyjob config,  there is a namespace for overall studyjob, a namespace for the workerspec, namespace for ConfigMap, namespace for metric collector . Won't this be complicated for the users by exposing much details? WDYT?

**Comment by hougangliu** [comment]
will submit another PR for v1alpha2

**Pull request closed without merging**